### PR TITLE
feat:  introduce ElectionClient trait for meta

### PR DESCRIPTION
**Pull request by shanicky**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

**This section will be used as the commit message. Please do not leave this empty!**

This PR introduces ElectionClient, where `run_once` executes an available campaign, which  will keep blocking if it runs normally, and when it returns normally, it represents a leader lost.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Installation and deployment 
* Connector (sources & sinks)
* SQL commands, functions, and operators
* RisingWave cluster configuration changes
* Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

### CODE REVIEW

**Comment by wangrunji0408** [comment]
Blocked by missing APIs in madsim etcd simulator. I'll fix them soon.

**Comment by shanicky** [comment]
This PR already works, but it will make the meta start a few seconds longer, which will cause some problems with deterministic testing, I'm still working on it, and it will be ready when it's solved.

**Comment by wangrunji0408** [comment]
Related fixes on etcd simulator:
- https://github.com/madsim-rs/madsim/pull/98
- https://github.com/madsim-rs/madsim/pull/101
- https://github.com/madsim-rs/madsim/pull/102

**Comment by CAJan93** [comment]
Hey @shanicky, let me know when this PR is ready for review. I think this is really important work and I would like to support you with it :)

**Review by yezizp2012 — COMMENTED** [review]
Great job!!! Generally LGTM.

File: src/meta/src/rpc/server.rs [hunk 1]
```diff
@@ -106,583 +120,89 @@ pub async fn rpc_serve(
     }
 }
 
-fn node_is_leader(leader_rx: &WatchReceiver<(MetaLeaderInfo, bool)>) -> bool {
-    leader_rx.borrow().clone().1
-}
-
 pub async fn rpc_serve_with_store<S: MetaStore>(
     meta_store: Arc<S>,
+    election_client: Option<ElectionClientRef>,
```
[the remark this answers is not in the document]
Comment (shanicky) [reply src/meta/src/rpc/server.rs:125]: It should be ok, but it is still semantically strange to implement an Election client for a standalone client, if it is a standalone deployment theoretically there should be no election process

**Review by yezizp2012 — APPROVED** [review]

**Comment by jon-chuang** [comment]
Why did this PR delete all of the tests introduced in https://github.com/risingwavelabs/risingwave/pull/6937

**Comment by shanicky** [comment]
> Why did this PR delete all of the tests introduced in https://github.com/risingwavelabs/risingwave/pull/6937?
> 
> 
> 
> Shouldn't we add back equivalent tests for those deleted? @shanicky

The previous test was used for the original kv mode election and had some problems, so it was removed, this pr introduces a simple madsim-based simulation test, and more simulation tests will be added after the meta client is complete

**Comment by shanicky** [comment]
> Hey @shanicky, 
> 
> I agree that using the simulation test is the superior approach, however should we maybe add/keep unit tests as well? Those may make it simpler to discover issues straight away. Sometimes finding the reason why a sim test failed is quite hard. 

The simulation test is easy to reproduce the problem because of madsim

**Pull request merged**